File: 395.Longest-Substring-with-At-Least-K-Repeating-Characters.go
// https://leetcode.com/problems/longest-substring-with-at-least-k-repeating-characters/description/
//
// algorithms
// Medium (36.85%)
// Total Accepted:    35.3K
// Total Submissions: 95.8K
// beats 100.0% of golang submissions
// 这道题主要在于找到一定满足不了要求的节点，然后将s切割成一个个子串进行递归

package leetcode

import "sort"

func longestSubstring(s string, k int) int {
	length := len(s)
	if length < k {
		return 0
	}

	numMap := make(map[rune]int)
	arrMap := make(map[rune][]int)
	for idx, ch := range s {
		if v, ok := numMap[ch]; ok {
			numMap[ch] = v + 1
			arrMap[ch] = append(arrMap[ch], idx)
		} else {
			numMap[ch] = 1
			arrMap[ch] = []int{idx}
		}
	}

	idxArr := []int{-1}
	for ch, v := range numMap {
		if v < k {
			idxArr = append(idxArr, arrMap[ch]...)
		}
	}
	notMatchLen := len(idxArr)
	if notMatchLen == 1 {
		return length
	}
	idxArr = append(idxArr, length)
	sort.Sort(sort.IntSlice(idxArr))

	res := 0
	for i := 0; i < notMatchLen; i++ {
		tmp := longestSubstring(s[idxArr[i]+1:idxArr[i+1]], k)
		if tmp > res {
			res = tmp
		}
	}

	return res
}

File: 395.Longest-Substring-with-At-Least-K-Repeating-Characters.py
# https://leetcode.com/problems/longest-substring-with-at-least-k-repeating-characters/
#
# algorithms
# Medium (38.36%)
# Total Accepted:    44,940
# Total Submissions: 117,151


from collections import defaultdict


class Solution(object):
    def longestSubstring(self, s, k):
        """
        :type s: str
        :type k: int
        :rtype: int
        """
        hash_map = defaultdict(int)
        for ch in s:
            hash_map[ch] += 1

        not_match_chs = set()
        for key, v in hash_map.iteritems():
            if v < k:
                not_match_chs.add(key)

        if not not_match_chs:
            return len(s)

        l = 0
        res = 0
        for idx, ch in enumerate(s):
            if ch in not_match_chs:
                res = max(res, self.longestSubstring(s[l:idx], k))
                l = idx + 1

        res = max(res, self.longestSubstring(s[l:], k))

        return res
